### DISCUSSION TITLE
Performance of WASM-based Plugins

### DISCUSSION THREAD

**Discussion by stahlbauer** [swc-project/swc#7521] [Q&A] [answered]
I have written a SWC plugin that is compiled into WebAssembly and then configured as (SWC) plugin from an NodeJs application. The performance is currently behind the expectations we had. The Babel version is about twice as fast (but consumes four time as much memory). 

Did you have a similar experience? Is there also a way to write "native" SWC plugins?

**Comment by stahlbauer** [comment]
Some more details: I am processing files that are in the range of megabytes, for example, adding coverage similar to the [swc-plugin-coverage-instrument](https://github.com/kwonoj/swc-plugin-coverage-instrument) plugin.

**Comment by kwonoj** [comment, marked as answer]
No for both. First, if you run swc-coverage-instrument test case itself, wasm plugin vs native transform performs nearly similar. When we measured back in the day we expects wasm plugin performs roughly 70-90% to the native codes at the best perf, depends on what it performs underlying. 

There is no such thing like native plugin. There was an experiments around those before introducing wasm plugin, but we dropped those idea due to it makes nearly impossible to plugin author to provide binaries for the each targeting platform. The only way is you build your own custom variant of swc as same as next.js does.

**Comment by kdy1** [comment]
Can you profile it?

**Comment by kwonoj** [comment]
If you can find bottleneck specific to swc, happy to look into it though.

**Reply by kwonoj** [reply]
Reading this again, 

>  range of megabytes

this _may_ affect some perf aspect as we as there's a cost to copy whole bytes into plugin's memory space. That's something we are aware and trying to improve, but afaik it isn't huge blockers for the most of general usecases. Still, not saying this is the root cause. There may be various reasons we can't say for sure.

**Comment by stahlbauer** [comment]
Hi. The WASM transpilation does not seem to be the problem (I would guess). Running the `fixture` test shows a similar performance . Which approach would you recommend for profiling test runs that rely on `testing::fixture`? Would `https://github.com/kdy1/cargo-profile` be applicable somehow? Thanks.

**Reply by stahlbauer** [reply]
@kdy1
